typing/btype.ml

```diff
@@ -496,25 +499,31 @@ let rec norm_univar ty =
   | Ttuple (ty :: _)   -> norm_univar ty
   | _                  -> assert false
 
-let rec copy_type_desc ?(keep_names=false) f = function
+let rec copy_type_desc ?(keep_names=false) f id_pairs = function
     Tvar _ as ty        -> if keep_names then ty else Tvar None
   | Tarrow (p, ty1, ty2, c)-> Tarrow (p, f ty1, f ty2, copy_commu c)
   | Ttuple l            -> Ttuple (List.map f l)
-  | Tconstr (p, l, _)   -> Tconstr (p, List.map f l, ref Mnil)
-  | Tobject(ty, {contents = Some (p, tl)})
-                        -> Tobject (f ty, ref (Some(p, List.map f tl)))
+  | Tconstr (p, l, _)   ->
+      let p' = Path.unsubst id_pairs (Path.subst id_pairs p) in
+      Tconstr (p', List.map f l, ref Mnil)
+  | Tobject(ty, {contents = Some (p, tl)}) ->
+      let p' = Path.unsubst id_pairs (Path.subst id_pairs p) in
+      Tobject (f ty, ref (Some(p', List.map f tl)))
   | Tobject (ty, _)     -> Tobject (f ty, ref None)
   | Tvariant _          -> assert false (* too ambiguous *)
   | Tfield (p, k, ty1, ty2) -> (* the kind is kept shared *)
       Tfield (p, field_kind_repr k, f ty1, f ty2)
   | Tnil                -> Tnil
-  | Tlink ty            -> copy_type_desc f ty.desc
+  | Tlink ty            -> copy_type_desc f id_pairs ty.desc
   | Tsubst _            -> assert false
   | Tunivar _ as ty     -> ty (* always keep the name *)
   | Tpoly (ty, tyl)     ->
       let tyl = List.map (fun x -> norm_univar (f x)) tyl in
       Tpoly (f ty, tyl)
-  | Tpackage (p, n, l)  -> Tpackage (p, n, List.map f l)
+  | Tpackage (p, n, l)  ->
+      let p' = Path.unsubst id_pairs (Path.subst id_pairs p) in
+      Tpackage (p', n, List.map f l)
+  | Tfunctor _          -> assert false (* must be handled by caller *)
```

Comment:
I think that's true of many of the cases of this function, yet we still handle them normally.
Is there any reason not to do that for this construction?

Reply:
Point taken.